### PR TITLE
Rewrite the dispatcher module

### PR DESCRIPTION
**Pull request by kidanger**
This PR is relatively large but simplifies a lot the code.
The class Dispatcher is now used directly by the core module. Core exposes the endpoint `get_demorunners_stats` for the CP2.
Typing annotations are added and most try/except blocks are removed. Error management is simplified using [result](https://pypi.org/project/result/).

Unit tests are cleaned-up, more exhaustive and reimplemented using pytest

Removed endpoints:
- /api/dispatcher/index, default, ping, shutdown
- /api/dispatcher/refresh_demorunners
- /api/dispatcher/set_policy

The last two were removed because I think it's better to have a system that is as much as possible declarative and not imperative. The dispatcher's policy is now set from core's config file.

### CODE REVIEW

File: ipol_demo/modules/dispatcher/test.py [hunk 1]
```diff
@@ -47,22 +47,13 @@ class DispatcherTests(unittest.TestCase):
...
-            status = response.get('status')
-        finally:
-            self.assertEqual(status, 'OK')
-
     def test_find_suitable_demorunner(self):
         """
         Test get suitable demorunner
         """
+        if True:
+            # test broken since the endpoint get_suitable_demorunner was removed
```
Comment (hmaciasc) [inline ipol_demo/modules/dispatcher/test.py]: What is preventing to remove the test now?

Comment (kidanger) [reply ipol_demo/modules/dispatcher/test.py]: Tests have been reimplemented in commit cbd91cf99f38a5edd0553a83859df3c23a0cf667.
Sorry for the large PR, I think it's good to review it both by commit and globally.

Comment (kidanger) [reply ipol_demo/modules/dispatcher/test.py]: I can rewrite the history to remove this "if True" in order to keep the commits clean.

Comment (hmaciasc) [reply ipol_demo/modules/dispatcher/test.py]: I saw later about the other commits, looks good to me 👍🏼

**Review by hmaciasc — APPROVED** [review]

**Comment by mcolom** [comment]
Before approving these changes, I have some questions:
- The Dispatcher's test has been removed from ci_tests/all.py, but we seems to have it at ipol_demo/modules/dispatcher/test.py since there are changes at that file. Why the Dispatcher's test was removed from the list at the first file?
- The service's URL has moved from /api/dispatcher/get_demorunners_stats to /api/core/get_demorunners_stats. I don't think we should change that URL simply because the code was moved. The API is not reflecting the structure of the system, but an independent interface. It's better to keep the old URL.
- In ipol_demo/modules/core/core.py we have at line 51 an import using sys.path.append... I guess this is because you're keeping the folder's structure and it's the only way you have to import code from there, since you can't do it with an standard Python's import. I don't think this is complete. If you move the Dispatcher's to Core, the right way would be to actually move it into Core's folder as a Python module and import it normally. Otherwise it's a workaround.

**Comment by kidanger** [comment]
Thank you for the review.
1. The new tests are using pytest and should be ran with `pytest ipol_demo/modules/dispatcher/test.py`. Right now `ci_tests/` is dedicated to run integration tests against each service. The new tests for dispatcher are more meant to be ran to validate individual commits / PR. I can set-up a first github actions workflow to run the dispatcher tests from there, to avoid the blind spot that my removal of dispatcher in all.py creates.
2. I agree; I'll take a look at cherrypy to see how we can have both URLs served by a single server.
3. OK. I'll can move the files to the core folder. Once most modules are converted, we can think about clean-up the folder tree a bit.

**Comment by mcolom** [comment]
Thanks for the changes! You can merge it into devel when you wish.

**Review by mcolom — APPROVED** [review]

**Comment by kidanger** [comment]
TODO when deploying on prod:
- update core's venv
- update nginx config (/api/dispatcher -> proxy to 8080 without rewrite)

**Pull request merged**